### PR TITLE
Fix dependencies

### LINKED ISSUE #125: Consolidate Requirements
We should consolidate libraries, gis, heroku, and vcs to requirements. We should also probably see if we can't trim any libraries we currently install (as it can take a fairly long time). May depend on switching to postgres in #121.

### PR DESCRIPTION
**Pull request by SillyInventor**
Resolves #125 

This pull merges all the dependencies into requirements, updates those requirements to latest, and fixes them all.

Require.io will be turned on once this pull is made, and all future dependency updates will be made via pull request.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -14,3 +14,5 @@ local_settings.py
 # Vagrant stuff
 .vagrant
 *.log
+
+.coverage
```
Comment (chuck211991) [inline .gitignore]: Technically, this is unrelated to this change. I'm fine with including it, because it's so minor.

File: activate [hunk 2]
```diff
@@ -1,9 +1,7 @@
 if [ ! -d "./python/" ]; then
     virtualenv python/
     . ./python/bin/activate
-    sudo pip install --upgrade -r libraries-vcs.txt
-    sudo pip install -r libraries.txt.lock
-    sudo pip install -r libraries-heroku.txt
+    sudo pip install --upgrade -r requirements.txt
 fi
 
 . ./python/bin/activate
```
Comment (chuck211991) [inline activate]: This entire script should probably be dropped if we are moving towards a Vagrant development; if you still want to maintain the ability to run on multiple systems, I suggest moving these steps to a Wiki page. Part of the reason for this is that I have about 3 different versions of this file; Arch Linux needs virtualenv2 to be called, windows doesn't understand the file at all, and most other systems use this version. If we move it to a Wiki with explanations, people can adapt it as needed.

Maybe another commit though?

File: requirements.txt [hunk 3]
```diff
@@ -1,40 +1,60 @@
...
+six==1.10.0
+geopy==1.11.0
+git+https://github.com/GK-12/SGIS-backend.git
 git+https://github.com/bartTC/django-attachments.git
-git+https://github.com/chuck211991/django-likes.git
+git+https://github.com/CSnap/django-likes.git
 git+https://github.com/chuck211991/django_teams.git
-git+https://github.com/chuck211991/django-cms-bootstrap-templates.git
+git+https://github.com/CSnap/django-templatetag-sugar.git
+git+https://github.com/divio/django-filer.git
```
Comment (chuck211991) [inline requirements.txt]: Do we want to keep dependencies that aren't explicitly used by us? For example, we don't use numpy, but include it because (I think) SGIS-backend uses it. If SGIS stops using it, we are now including it for no reason (and this particular example is important, because it takes numpy like 15 minutes to compile on my Windows install)

We should (and never have, shame on us) keep track of WHY we include each library, either in this file (a simple one line comment would suffice) or in a requirements document somewhere (in this context, requirements means list of features Ron wants). This would be further reason for removing dependencies pulled in by other libraries.

File: requirements.txt [hunk 4]
```diff
@@ -1,40 +1,60 @@
...
-django-extra-views>=0.6.2
-djangorestframework>=2.3.12
-django-allauth>=0.18.0
-six>=1.3.0
-wsgiref>=0.1.2
-python-memcached>=1.53
-sorl-thumbnail>=11.12.1b
-django-taggit-templatetags>=0.4.6dev
-django-extensions>=1.3.3
+bbcode==1.0.24
```
Comment (chuck211991) [inline requirements.txt]: This file should be requirements.txt.lock; we need to note that it locks versions, even if you think those are going to be automatically updated.

File: requirements.txt [hunk 5]
```diff
@@ -1,40 +1,60 @@
...
+oauthlib==2.0.1
+Pillow==4.0.0
+psycopg2==2.7
+pybbm==0.18.4
+pytz==2016.10
+PyYAML==3.12
+requests-oauthlib==0.8.0
+six==1.10.0
+geopy==1.11.0
+git+https://github.com/GK-12/SGIS-backend.git
```
Comment (chuck211991) [inline requirements.txt]: We should see if we can move these git+ libraries into pypi; this solution was meant to be temporary. Anything that never makes it into pypi is hard to track because we can't rely on the Git repo remaining intact for long periods of time, and we may want to consider dropping.

File: requirements.txt [hunk 6]
```diff
@@ -1,40 +1,60 @@
...
+numpy==1.12.0
+oauthlib==2.0.1
+Pillow==4.0.0
+psycopg2==2.7
+pybbm==0.18.4
+pytz==2016.10
+PyYAML==3.12
+requests-oauthlib==0.8.0
+six==1.10.0
+geopy==1.11.0
```
Comment (chuck211991) [inline requirements.txt]: This should be alphabetized

File: .travis.yml [hunk 7]
```diff
@@ -17,10 +17,8 @@ install:
...
   - sudo $(type -P npm) install -g yuglify
   - sudo pip install setuptools --upgrade
-  - sudo pip install --upgrade -r libraries-vcs.txt
-  - sudo pip install -r libraries.txt
-  - sudo pip install -r libraries-heroku.txt
-
+  - sudo pip install --upgrade -r requirements.txt
+  
 before_script:
   - psql -c "CREATE DATABASE travisci;" -U postgres
```
Comment (chuck211991) [inline .travis.yml]: Unrelated to this commit, but this file should somehow rely on the setup script from the Vagrant install; or maybe we should move the database part of the setup script out, and have both this and Vagrant use that. DRY

File: .gitignore (hunk 1 above)
Comment (SillyInventor) [inline .gitignore]: My justification was that because coverage is in requirements, it's fine to leave there, but installing and running tests can add said file... so it's related, but not very. The other option was taking it out and changing it in another commit. But we want it anyways so...

File: activate (hunk 2 above)
Comment (SillyInventor) [inline activate]: I don't want to do that here.

I am also, still, wary of breaking other people's systems. I know you think everyone should drop what they're doing to do it the new way, but it's just not necessary IMO.

File: requirements.txt (hunk 4 above)
Comment (SillyInventor) [inline requirements.txt]: I see no need to add .lock, it is fairly non-standard to add .lock (I haven't seen any other examples of django apps doing this). Furthermore, requires.io looks for .txt files, so we'd have to setup a different method.

File: activate (hunk 2 above)
Comment (chuck211991) [inline activate]: Fair enough, we can leave it for now. But eventually someday we should come up with a more permanent solution.

File: requirements.txt (hunk 5 above)
Comment (SillyInventor) [inline requirements.txt]: I don't want to make this change in this pull.

Maybe I'm confused, but I don't see why we can't rely on the git repo remaining intact and/or are unable to track changes.

File: requirements.txt (hunk 3 above)
Comment (SillyInventor) [inline requirements.txt]: I think we need to start treating GIS as being part of the site. The site handles urls to GIS, there are calls in settings for GIS, and when GIS goes down people see it as part of the site going down. Besides, if the undergraduates follow through and make it accessible from the applications page, then we're going to need to maintain it the same.

File: requirements.txt (hunk 4 above)
Comment (chuck211991) [inline requirements.txt]: It's fairly common when deploying an application to lock the dependencies; it's possible that many Django apps don't publicize this information (since it really shouldn't be public, in case there's a security flaw in one of the dependencies), but you can see it in other web projects/frameworks (such as Ruby, Phoenix, Node, anything that uses Maven, etc.). There is also the possibility that many apps we look at haven't been in production for long enough to have this issue; and this has been an issue for us. It has been an issue many times. I personally did not do this until we started have all the problems with it.

We could call it requirements.lock.txt, but IMO, it seems like a very flimsy system if you can't specify the requirements file :-P. If it ends up being a show stopper, and you really want to use that site, go ahead and ignore this advice, but I reserve the right to "I told you so".

File: requirements.txt (hunk 5 above)
Comment (chuck211991) [inline requirements.txt]: So a huge part of requirements engineering is that version 0.10.12 should ALWAYS refer to same bit of code. When we use git repos like this, that is not the case. It happens all the time that someone (*cough*me*cough*) forgets to update the version number when a change is made and pushed to Git; so now we can not easily revert months later. I believe that pypi doesn't let that shenanigans happen.

Ok if we hold off on it, but if you actually want to FIX the dependencies, rather than just moving files around, this is a huge part.

File: requirements.txt (hunk 3 above)
Comment (chuck211991) [inline requirements.txt]: Sure, but do we want to keep dependencies that aren't explicitly used by us? The issue could easily happen with django-filer; it depends on django-mptt, then it decides django-mpttis a pain, gets rid of it, now we include django-mptt for no reason.

Comment (chuck211991) [inline requirements.txt]: And still, the why we include each library.

File: .travis.yml (hunk 7 above)
Comment (SillyInventor) [inline .travis.yml]: If you want you can make an issue & pull for that.

File: requirements.txt (hunk 4 above)
Comment (chuck211991) [inline requirements.txt]: After looking at https://requires.io/, I think the best approach would be to name this file requirements.t.xt.lock, and setup a setup.py which includes it. This way we can maintain the list of libraries, and the why, in addition to locked versions that will get overwritten by requires.io everytime there is a new version.

File: requirements.txt [hunk 8]
```diff
@@ -1,40 +1,60 @@
...
-six>=1.3.0
-wsgiref>=0.1.2
-python-memcached>=1.53
-sorl-thumbnail>=11.12.1b
-django-taggit-templatetags>=0.4.6dev
-django-extensions>=1.3.3
+bbcode==1.0.24
+BeautifulSoup==3.2.1
+coverage==4.3.4
+Django==1.10.6
```
Comment (chuck211991) [inline requirements.txt]: We don't currently use Django 1.10; did you check to make sure everything works?

File: requirements.txt (hunk 5 above)
Comment (SillyInventor) [inline requirements.txt]: I am not sure I agree.

The fix in the branch name was meant to use the second definition: make specific and determined... adding constraints.

File: requirements.txt (hunk 6 above)
Comment (SillyInventor) [inline requirements.txt]: Thanks, should be alphabetized now.

File: requirements.txt (hunk 8 above)
Comment (SillyInventor) [inline requirements.txt]: I did, but on a second more thorough check (prompted by your question) found a bug. I will fix it tomorrow.

**Comment by chuck211991** [comment]
FYI, this change updates a few minor versions and adds one library (SGIS-backend) when compared to the old libraries.txt.lock (which used a ~=, so the minor versions would already be updated). Here's a list of major changes that should be looked into:

- Django==1.10.6
- django-bootstrap3==8.2.1
- django-compressor==2.1.1
- Django-Select2==5.8.10 (what do we use this for?)
- djangorestframework==3.5.4
- djangorestframework-csv==2.0.0
- Pillow==4.0.0

The rest of the changes are likely inconsequential from the old libraries.txt.lock file. These changes *might* result in broken functionality; especially django 1.10, djangorestframework, and djangorestframework-csv. We can't easily test djangorestframework-csv until https://github.com/CSnap/rpi_csdt_community/issues/129 is resolved.

**Comment by SillyInventor** [comment]
Although it builds well, and pages on the default don't break, there are some conflicts with Django 1.10. I am working on it.

**Comment by SillyInventor** [comment]
@chuck211991, I fixed all the bugs I have found.

File: project_share/models.py [hunk 9]
```diff
@@ -48,13 +49,20 @@ def __unicode__(self):
 
 class Approval(models.Model):
     project = models.OneToOneField('Project')
-    when_requested = models.DateTimeField(auto_now_add=True, null=True, blank=True)
```
Comment (chuck211991) [inline project_share/models.py]: Why was auto_now_add=True dropped? Don't we need this? It allows us to audit stuff if there's ever a problem; otherwise, we may as well ditch (at least one) this and when_updated

File: rpi_csdt_community/views.py [hunk 10]
```diff
@@ -9,7 +9,7 @@
 
 def comment_post_wrapper(request):
     # Clean the request to prevent form spoofing
-    if request.user.is_authenticated():
+    if request.user.is_authenticated:
```
Comment (chuck211991) [inline rpi_csdt_community/views.py]: I don't think we support comments on anything anymore; maybe this code should just be dropped?

File: project_share/models.py (hunk 9 above)
Comment (chuck211991) [inline project_share/models.py]: Oh, I see, you added code further down to replace this. But why?

Comment (SillyInventor) [inline project_share/models.py]: Because Django was treating auto_now badly, and on further inspection the internet indicated that use of auto_now is considered bad form and not appropriate, and instead to use the save method. Replacing that remove bugs in the saving and approval process.

File: rpi_csdt_community/views.py (hunk 10 above)
Comment (SillyInventor) [inline rpi_csdt_community/views.py]: ... maybe... but not here.

**Comment by SillyInventor** [comment]
Sure, I can switch us back to old dependencies and do this update separately.

File: project_share/models.py (hunk 9 above)
Comment (chuck211991) [inline project_share/models.py]: Could you elaborate on "treating auto_now badly" and "the internet indicated..."? I generally don't like rewriting code like this unless there is a very persuasive argument.

Comment (SillyInventor) [inline project_share/models.py]: I was getting errors on saving projects and approvals in Django 10, as it was creating the datetimefield without a time first, and then complaining that it was null when it's non-nullable.

Comment (chuck211991) [inline project_share/models.py]: It's still listed in the Django docs, so if we can't find something else that is causing this error, we should report it (https://docs.djangoproject.com/en/1.10/ref/models/fields/#datefield). Like I said though, I would really like to document a convincing argument for re-implementing this logic in the save function.

**Comment by chuck211991** [comment]
Thanks, I would greatly appreciate that.

File: requirements.txt (hunk 4 above)
Comment (SillyInventor) [inline requirements.txt]: I know it's in documentation. but on inspection one finds [this stackoverflow comment](http://stackoverflow.com/questions/1737017/django-auto-now-and-auto-now-add), which clarifies that it breaks depending on your Django version and this seems to be the case here. Particularly, it seems to be clashing with not null which I think is imposed by teams.

**Comment by SillyInventor** [comment]
Reverted.

**Comment by chuck211991** [comment]
Ok, this commit looks fine. Please verify that a "vagrant destroy && vagrant up" still work as expected, then go ahead and commit.

**Comment by SillyInventor** [comment]
I always test fresh before a commit. @chuck211991 you shouldn't approve things you haven't tested. @ryaholm please test this code and #119 on Monday and make the pull if you find no errors.

**Comment by chuck211991** [comment]
Worked fine for me, mostly. Had to add a dependency to the installation file for Vagrant. Pull incoming...
